### ISSUE TITLE
time_bookings "Internal error"

### ISSUE THREAD

**Issue by yzenezy** [hicknhack-software/redmine_hourglass#144]
Hi,

I've migrated from from Redmine 3.4.5.stable to Redmine 4.2.1.stable. We previously used redmine_hourglass but I didn't see a version listed as compatible with Redmiine 4.2.1 so I have installed the latest developement branch of redmine_hourglass.

I managed to get the development branch mostly working except for time_bookings. I get an "internal error" when browsing to https://myredmine/hourglass/ui/time_bookings. The log shows the following:

```
#Started GET "/hourglass/ui/time_bookings" for 127.0.0.1 at 2021-05-18 07:51:11 +1000
#Processing by HourglassUiController#time_bookings as HTML
#  Current user: tom (id=20)
#  Rendering plugins/redmine_hourglass/app/views/hourglass_ui/time_bookings.slim within layouts/base
#  Rendered plugins/redmine_hourglass/app/views/hourglass_ui/lists/_chart.slim (73.6ms)
#  Rendered plugins/redmine_hourglass/app/views/hourglass_ui/time_bookings.slim within layouts/base (73.8ms)
#Completed 500 Internal Server Error in 336ms (ActiveRecord: 86.5ms)
#  
#ActionView::Template::Error (Mysql2::Error: Unknown column 'date' in 'field list': SELECT 
SUM(time_entries.hours) AS sum_time_entries_hours, date AS date, time_entries.project_id AS 
time_entries_project_id FROM `hourglass_time_bookings` INNER JOIN `time_entries` ON `time_entries`.`id` = 
`hourglass_time_bookings`.`time_entry_id` INNER JOIN `projects` ON `projects`.`id` = `time_entries`.`project_id`  
LEFT OUTER JOIN `time_entries` `time_entries_hourglass_time_bookings` ON 
`time_entries_hourglass_time_bookings`.`id` = `hourglass_time_bookings`.`time_entry_id` LEFT OUTER JOIN 
`time_entries` `time_entries_hourglass_time_bookings_join` ON 
`time_entries_hourglass_time_bookings_join`.`id` = `hourglass_time_bookings`.`time_entry_id` LEFT OUTER 
JOIN `enumerations` ON `enumerations`.`id` = `time_entries_hourglass_time_bookings_join`.`activity_id` AND 
`enumerations`.`type` IN ('TimeEntryActivity') LEFT OUTER JOIN `hourglass_time_logs` ON 
`hourglass_time_logs`.`id` = `hourglass_time_bookings`.`time_log_id` LEFT OUTER JOIN `users` ON `users`.`id` 
= `hourglass_time_logs`.`user_id` AND `users`.`type` IN ('User', 'AnonymousUser') LEFT OUTER JOIN 
`time_entries` `time_entries_hourglass_time_bookings_join_3` ON 
`time_entries_hourglass_time_bookings_join_3`.`id` = `hourglass_time_bookings`.`time_entry_id` LEFT OUTER 
JOIN `issues` ON `issues`.`id` = `time_entries_hourglass_time_bookings_join_3`.`issue_id` LEFT OUTER JOIN 
`versions` ON `versions`.`id` = `issues`.`fixed_version_id` WHERE (hourglass_time_bookings.start > '2021-04-30 
23:59:59.999999' AND hourglass_time_bookings.start <= '2021-05-31 23:59:59.999999') AND `projects`.`id` IN 
(6, 42, 196, 197, 198, 199, 200, 201, 203, 204, 205, 208, 210, 213, 217, 218, 234) GROUP BY date, 
time_entries.project_id):
#    1: - data, ticks, tooltips = chart_data @chart_query
#    2: javascript:
#    3:   window.hourglass = window.hourglass || {};
#    4:   window.hourglass.jqplotData = {
#  
#app/models/query.rb:1080:in `rescue in grouped_query'
#app/models/query.rb:1069:in `grouped_query'
#app/models/query.rb:1027:in `total_by_group_for'
#plugins/redmine_hourglass/app/helpers/hourglass/chart_helper.rb:37:in `hours_per_date'
#plugins/redmine_hourglass/app/helpers/hourglass/chart_helper.rb:11:in `chart_data'
#plugins/redmine_hourglass/app/views/hourglass_ui/lists/_chart.slim:1:in 
`_plugins_redmine_hourglass_app_views_hourglass_ui_lists__chart_slim__3952624100409346438_69857937878080'
#plugins/redmine_hourglass/app/views/hourglass_ui/time_bookings.slim:2:in `block in _plugins_redmine_hourglass_app_views_hourglass_ui_time_bookings_slim__3659909628173303537_69857937741360'
#plugins/redmine_hourglass/app/views/hourglass_ui/time_bookings.slim:1:in 
`_plugins_redmine_hourglass_app_views_hourglass_ui_time_bookings_slim__3659909628173303537_69857937741360'
#lib/redmine/sudo_mode.rb:61:in `sudo_mode'
```
 
During install and configuration I had to modify the Gemfile as there was an issue with a locked version of sprockets 4.0.2. Then had to add sassc-rails and remove rqrcode. That seemed to work with bundle install and precompiling the assets. The  differences in the Gemfile are shown below:

```
# diff redmine_hourglass_orig/Gemfile /usr/share/webapps/redmine/plugins/redmine_hourglass/Gemfile
7c7,9
< gem 'sprockets', '~> 3.7.2', require: 'sprockets/railtie'
---
> #gem 'sprockets', '~> 3.7.2', require: 'sprockets/railtie'
> gem 'sprockets', '~> 4.0.2', require: 'sprockets/railtie'
> gem 'sassc-rails'
23c25
< gem 'rqrcode', '~> 0.10.1'
---
> #gem 'rqrcode', '~> 0.10.1'
```
Other than that I haven't made any other changes.

Have I broken something with my Gemfile changes or is there something else need to get the time_bookings working again?

Can someone please comment/help with the "internal error" message above?

Kind regards,
Tom

**Comment by arBmind** [comment]
Hi @yzenezy,
thank you for your efforts to try using hourglass with latest Redmine version. Unfortunately we didn't have time to try this version.

From the first look it seems like the "date" column is missing or was renamed, which is an essential part of the broken query. It does not state which model, but I guess it was on time_entries (provided by Redmine itself).

All our team members are currently busy. Feel free to investigate further. PR is always welcome.

**Comment by arBmind** [comment]
It seems the Query model has changed. This requires us to use the proper `TimestampQueryColumn` to prevent this misbehaviour.

#146 fixes this.

**Comment by yzenezy** [comment]
@arBmind Thanks for the update! This latest change really helps.

I've deployed the development code and can now browse to https://myredmine/hourglass/ui/time_bookings.

For my installation I still had to munge the Gemfile to do the migration and assets build:

```
# diff Gemfile ../../share/webapps/redmine/plugins/redmine_hourglass/Gemfile
7c7,8
< gem 'sprockets', '~> 3.7.2', require: 'sprockets/railtie'
---
> gem 'sprockets', '~> 4.0.2', require: 'sprockets/railtie'
> gem 'sassc-rails'
```

Unfortunately I can't apply all the Time Booking filters successfully. I get an **Internal Error** when selecting **Date >=**  or using a predefined Date like **this year**.

```
I, [2021-06-24T09:25:34.129882 #2556]  INFO -- : Started GET "/hourglass/ui/time_bookings?utf8=%E2%9C%93&set_filter=1&f%5B%5D=date&op%5Bdate%5D=%3E%3D&v%5Bdate%5D%5B%5D=2020-12-24&f%5B%5D=&c%5B%5D=date&c%5B%5D=start&c%5B%5D=stop&c%5B%5D=hours&c%5B%5D=project&c%5B%5D=issue&c%5B%5D=activity&c%5B%5D=comments&group_by=&t%5B%5D=&t%5B%5D=hours&t%5B%5D=" for 127.0.0.1 at 2021-06-24 09:25:34 +1000
I, [2021-06-24T09:25:34.132313 #2556]  INFO -- : Processing by HourglassUiController#time_bookings as HTML
I, [2021-06-24T09:25:34.132408 #2556]  INFO -- :   Parameters: {"utf8"=>"✓", "set_filter"=>"1", "f"=>["date", ""], "op"=>{"date"=>">="}, "v"=>{"date"=>["2020-12-24"]}, "c"=>["date", "start", "stop", "hours", "project", "issue", "activity", "comments"], "group_by"=>"", "t"=>["", "hours", ""]}
I, [2021-06-24T09:25:34.154075 #2556]  INFO -- :   Current user: tom (id=20)
I, [2021-06-24T09:25:34.433897 #2556]  INFO -- :   Rendering plugins/redmine_hourglass/app/views/hourglass_ui/time_bookings.slim within layouts/base
I, [2021-06-24T09:25:34.600554 #2556]  INFO -- :   Rendered plugins/redmine_hourglass/app/views/hourglass_ui/lists/_chart.slim (166.4ms)
I, [2021-06-24T09:25:34.600704 #2556]  INFO -- :   Rendered plugins/redmine_hourglass/app/views/hourglass_ui/time_bookings.slim within layouts/base (166.7ms)
I, [2021-06-24T09:25:34.600914 #2556]  INFO -- : Completed 500 Internal Server Error in 468ms (ActiveRecord: 95.5ms)
F, [2021-06-24T09:25:34.601724 #2556] FATAL -- :   
F, [2021-06-24T09:25:34.601779 #2556] FATAL -- : ActionView::Template::Error (can't iterate from NilClass):
F, [2021-06-24T09:25:34.601940 #2556] FATAL -- :     1: - data, ticks, tooltips = chart_data @chart_query
    2: javascript:
    3:   window.hourglass = window.hourglass || {};
    4:   window.hourglass.jqplotData = {
F, [2021-06-24T09:25:34.601994 #2556] FATAL -- :   
F, [2021-06-24T09:25:34.602035 #2556] FATAL -- : plugins/redmine_hourglass/app/helpers/hourglass/chart_helper.rb:17:in `block in chart_data'
plugins/redmine_hourglass/app/helpers/hourglass/chart_helper.rb:16:in `each'
plugins/redmine_hourglass/app/helpers/hourglass/chart_helper.rb:16:in `chart_data'
plugins/redmine_hourglass/app/views/hourglass_ui/lists/_chart.slim:1:in `_plugins_redmine_hourglass_app_views_hourglass_ui_lists__chart_slim__3402368483348434317_47069076477060'
plugins/redmine_hourglass/app/views/hourglass_ui/time_bookings.slim:2:in `block in _plugins_redmine_hourglass_app_views_hourglass_ui_time_bookings_slim__1055552208215350279_47069075724500'
plugins/redmine_hourglass/app/views/hourglass_ui/time_bookings.slim:1:in `_plugins_redmine_hourglass_app_views_hourglass_ui_time_bookings_slim__1055552208215350279_47069075724500'
lib/redmine/sudo_mode.rb:61:in `sudo_mode'

```
Some of the **Date** filters work without error but return no data (tomorrow, today, yesterday, next week, this week, last week, last 2 weeks, next month, this month, last month, days ago, none, etc.).

I was thinking that perhaps there is no data (and no error) on some of the **Date** filter values because I haven't been booking anything recently so I booked something from yesterday and applied the **Date** filter with values like **yesterday**, and **this week** and immediately got the **Internal error** again.

Is this related to my Gemfile munge or is there still something else in the code not compatible on Redmine 4.2.1?

Kind regards,
Tom

**Comment by arBmind** [comment]
@yzenezy Thank you for trying it out again.
The `date` filter is not actually the problem here. The stack trace indicates that the chart_data could not be retrieved. The chart groups your time_bookings by `date` to show a bar for each day.

The current development branch resolved this issue for me. I will take another look into it, as soon as possible.

**Comment by arBmind** [comment]
@yzenezy I got away with running `bundle update sprockets` to get sprockets back to version 3.7.2 and everything seems to work.

**Comment by yzenezy** [comment]
I managed to get sprockets 3.7.2 into this plugins bundle a couple of weeks ago (sorry not to report sooner; so much going on!). It was locked by the redmine application's Gemfile. When I got sprockets 3.7.2 working with Redmine itself, I was able to get it working for redmine_hourglass, too. I'm not really sure how that works (why Gems from redmine affect Gems from plugin). The redmine and it's environment was not set up by me so maybe I'm working with an overly complicated configuration...

Anyway, thanks for all your help.
